src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.cs

```diff
@@ -48,20 +52,38 @@ internal partial class WorkspaceProjectContextHost : AbstractMultiLifetimeCompon
             _activeConfiguredProjectProvider = activeConfiguredProjectProvider;
             _applyChangesToWorkspaceContextFactory = applyChangesToWorkspaceContextFactory;
             _dataProgressTrackerService = dataProgressTrackerService;
+            _languageServiceTelemetryService = languageServiceTelemetryService;
+        }
+
+        private string ProjectId
+        {
+            get
+            {
+                if (string.IsNullOrEmpty(_projectId))
+                {
+                    string? fullPath = _project?.UnconfiguredProject?.FullPath;
+                    _projectId = string.IsNullOrEmpty(fullPath) ? string.Empty : _languageServiceTelemetryService.HashValue(fullPath!);
+                }
+
+                return _projectId!;
```

Comment:
I think this would work to remove some `!`s:

```suggestion
                    _projectId = Strings.IsNullOrEmpty(fullPath) ? string.Empty : _languageServiceTelemetryService.HashValue(fullPath);
                }

                return _projectId;
```